### PR TITLE
Fix Python Gateway client trusts Gateway status addresses without validation

### PR DESCRIPTION
**Pull request by chw120**
#### What this PR does / why we need it:
This PR addresses a medium-risk security vulnerability (Finding 9) where the Python client for the agentic sandbox would directly accept and use addresses from a Gateway resource's status block without validation.

**Key changes included:**

- **Address Validation Helper**: Added _is_valid_ip (using the ipaddress module) and _is_valid_gateway_hostname to ensure that the retrieved address conforms to expected formats.
- **Special Character Filtering**: Implemented a check to reject any address containing special characters like /, ?, #, or @, which are commonly used in SSRF payloads to manipulate URL parsing.
- **Secure Gateway Resolution**: Updated the wait_for_gateway_ip method in both synchronous and asynchronous helpers to apply these validation checks before returning the address to the caller.
- **Unit Tests**: Added comprehensive tests in test_k8s_helper.py to verify that valid IPs and hostnames are accepted while malformed addresses (e.g., those with special characters or invalid formats) are correctly rejected.

#### Release Note
```
SECURITY: Fixed a vulnerability where the Python SDK blindly trusted addresses from Gateway resource statuses. This fix introduces strict validation for IP addresses and hostnames to prevent SSRF.
```

## Release Notes

* **Bug Fixes**
  * Improved gateway external-address selection by validating candidates before returning them, ignoring malformed/non-address entries, and choosing the first verified IPv4 or compliant hostname.
  * Added stricter hostname validation (label length/format) and improved handling of valid IPv6 addresses.
  * Kept support for “disguised” IPv4 inputs (decimal/hex and related variants) to resolve correctly.

* **Tests**
  * Expanded unit coverage for mixed watch events, multi-address lists, invalid encodings, hostname edge cases, and type-robustness (including IPv6).

### CODE REVIEW

File: clients/python/agentic-sandbox-client/k8s_agent_sandbox/async_k8s_helper.py [hunk 1]
```diff
@@ -260,6 +260,14 @@ async def list_sandbox_claims(self, namespace: str) -> list[str]:
             logger.error(f"Error listing sandbox claims in namespace {namespace}: {e}")
             raise
 
+    def _is_valid_ip(self, s: str) -> bool:
```
Comment (barney-s) [inline clients/python/agentic-sandbox-client/k8s_agent_sandbox/async_k8s_helper.py]: The private methods `_is_valid_ip` and `_is_valid_gateway_hostname` add unnecessary boilerplate and indirection. Additionally, performing local dynamic imports (`from .utils import ...`) inside these methods on every call introduces runtime overhead within the watch loop.

Since `.utils` has no dependencies on `async_k8s_helper`, there is no circular dependency risk. Consider importing `is_valid_ip` and `is_valid_gateway_hostname` at the top of the file and calling them directly in `wait_for_gateway_ip`.

File: clients/python/agentic-sandbox-client/k8s_agent_sandbox/async_k8s_helper.py [hunk 2]
```diff
@@ -287,11 +295,28 @@ async def wait_for_gateway_ip(self, gateway_name: str, namespace: str, timeout:
                         gateway_object = event["object"]
                         status = gateway_object.get("status") or {}
                         addresses = status.get("addresses", [])
-                        if addresses:
-                            ip_address = addresses[0].get("value")
-                            if ip_address:
-                                logger.info(f"Gateway ready. IP: {ip_address}")
-                                return ip_address
+                        for address in addresses:
```
Comment (barney-s) [inline clients/python/agentic-sandbox-client/k8s_agent_sandbox/async_k8s_helper.py:329]: Defensive check: It is safer to verify that `address` is a dictionary (e.g., `isinstance(address, dict)`) before calling `.get("value")`. This protects against any unexpected API response formats or schema updates that could otherwise throw an `AttributeError`.

File: clients/python/agentic-sandbox-client/k8s_agent_sandbox/async_k8s_helper.py [hunk 3]
```diff
@@ -287,11 +295,28 @@ async def wait_for_gateway_ip(self, gateway_name: str, namespace: str, timeout:
...
+                            ip_address = address.get("value")
+                            if not ip_address:
+                                continue
```
Comment (barney-s) [inline clients/python/agentic-sandbox-client/k8s_agent_sandbox/async_k8s_helper.py:334]: The denylist check `any(c in ip_address for c in "/?#@")` is redundant. Since both `is_valid_ip` and `is_valid_gateway_hostname` strictly validate against allowlist (valid IPv4 addresses and standard hostname characters), any string containing `/`, `?`, `#`, or `@` will be automatically rejected.

Strict allowlist validation is more robust and secure than mixing it with incomplete denylist checks.

File: clients/python/agentic-sandbox-client/k8s_agent_sandbox/k8s_helper.py [hunk 4]
```diff
@@ -216,6 +216,14 @@ def list_sandbox_claims(self, namespace: str) -> List[str]:
             logging.error(f"Error listing sandbox claims in namespace {namespace}: {e}")
             raise
 
+    def _is_valid_ip(self, s: str) -> bool:
```
Comment (barney-s) [inline clients/python/agentic-sandbox-client/k8s_agent_sandbox/k8s_helper.py]: The private methods `_is_valid_ip` and `_is_valid_gateway_hostname` add unnecessary boilerplate and indirection. Additionally, performing local dynamic imports (`from .utils import ...`) inside these methods on every call introduces runtime overhead within the watch loop.

Since `.utils` has no dependencies on `k8s_helper`, there is no circular dependency risk. Consider importing `is_valid_ip` and `is_valid_gateway_hostname` at the top of the file and calling them directly in `wait_for_gateway_ip`.

File: clients/python/agentic-sandbox-client/k8s_agent_sandbox/k8s_helper.py [hunk 5]
```diff
@@ -240,9 +248,26 @@ def wait_for_gateway_ip(self, gateway_name: str, namespace: str, timeout: int) -
                     gateway_object = event['object']
                     status = gateway_object.get('status') or {}
                     addresses = status.get('addresses', [])
-                    if addresses:
-                        ip_address = addresses[0].get('value')
-                        if ip_address:
-                            logging.info(f"Gateway ready. IP: {ip_address}")
-                            w.stop()
-                            return ip_address
+                    for address in addresses:
```
Comment (barney-s) [inline clients/python/agentic-sandbox-client/k8s_agent_sandbox/k8s_helper.py:293]: Defensive check: It is safer to verify that `address` is a dictionary (e.g., `isinstance(address, dict)`) before calling `.get("value")`. This protects against any unexpected API response formats or schema updates that could otherwise throw an `AttributeError`.

File: clients/python/agentic-sandbox-client/k8s_agent_sandbox/k8s_helper.py [hunk 6]
```diff
@@ -240,9 +248,26 @@ def wait_for_gateway_ip(self, gateway_name: str, namespace: str, timeout: int) -
...
+                        ip_address = address.get('value')
+                        if not ip_address:
+                            continue
```
Comment (barney-s) [inline clients/python/agentic-sandbox-client/k8s_agent_sandbox/k8s_helper.py:298]: The denylist check any(c in ip_address for c in "/?#@") is redundant. Since both is_valid_ip and is_valid_gateway_hostname strictly validate against allowlist (valid IPv4 addresses and standard hostname characters), any string containing /, ?, #, or @ will be automatically rejected.

Strict allowlist validation is more robust and secure than mixing it with incomplete denylist checks.

File: clients/python/agentic-sandbox-client/k8s_agent_sandbox/utils.py [hunk 7]
```diff
@@ -43,3 +43,33 @@ def construct_sandbox_claim_lifecycle_spec(shutdown_after_seconds: int) -> dict[
...
+        # Downstream callers construct URLs with `http://{ip_address}`,
+        # which only works for IPv4 literals without additional
+        # normalization. Reject IPv6 here to keep the returned value
+        # compatible with existing URL construction.
+        return isinstance(ipaddress.ip_address(s), ipaddress.IPv4Address)
+    except ValueError:
+        return False
+
+
+def is_valid_gateway_hostname(s: str) -> bool:
```
Comment (barney-s) [inline clients/python/agentic-sandbox-client/k8s_agent_sandbox/utils.py:123]: While the hostname validator correctly restricts the overall length to 253 characters, it does not check that individual DNS labels (separated by `.`) do not exceed the 63-character limit as specified by RFC 1123 / RFC 1035.

Consider adding a check such as `any(len(label) > 63 for label in s.split('.'))` to strictly enforce DNS specs.

**Review by barney-s — COMMENTED** [review]
- **Performance & Boilerplate**: Recommends removing unnecessary class helper methods that perform dynamic local imports on every call, in favor of top-level imports.
- **Redundant Validation**: Highlights a redundant blacklist character check that is already fully covered by the strict whitelists of the IP and hostname validation functions.
- **RFC Hostname Completeness**: Identifies a missing DNS label length check (63-character limit) in the hostname validator.

**Comment by SHRUTI6991** [comment]
can you fix the tests @chw120 ?

File: clients/python/agentic-sandbox-client/k8s_agent_sandbox/utils.py [hunk 8]
```diff
@@ -43,3 +43,33 @@ def construct_sandbox_claim_lifecycle_spec(shutdown_after_seconds: int) -> dict[
...
+    import ipaddress
+    try:
+        # Downstream callers construct URLs with `http://{ip_address}`,
+        # which only works for IPv4 literals without additional
+        # normalization. Reject IPv6 here to keep the returned value
+        # compatible with existing URL construction.
+        return isinstance(ipaddress.ip_address(s), ipaddress.IPv4Address)
+    except ValueError:
+        return False
+
```
Comment (aditya-shantanu) [inline clients/python/agentic-sandbox-client/k8s_agent_sandbox/utils.py:110]: `is_valid_gateway_hostname` accepts an all-digit string (e.g. `"2130706433"`): digits are in the allowlist and nothing requires an alphabetic component or a dot. Such a value also fails `is_valid_ip` (Pythons `ipaddress` rejects non-dotted integer forms), so it falls through the `is_valid_ip(...) or is_valid_gateway_hostname(...)` guard and is accepted. Downstream then builds `http://2130706433`, and `socket.gethostbyname("2130706433")` coerces the integer to `127.0.0.1` (libc `inet_aton` integer-form behavior). That is exactly the SSRF-to-loopback bypass this PR is meant to close: a compromised Gateway status could hand the client a decimal/integer-encoded address that redirects to loopback or other internal IPs while passing validation.

Consider rejecting bare numeric tokens that are not valid dotted IPv4 — e.g. reject when `s.replace(".", "").isdigit()` is true but `is_valid_ip(s)` is false, or require at least one alphabetic character in a hostname. Reasoning: this validator is the security boundary for addresses sourced from an untrusted status block, and integer-literal IPs are a well-known allowlist-bypass for SSRF.

(Heads-up: please apply any change locally rather than via the GitHub "Commit suggestion" button — that adds the bot as co-author and breaks the CLA check.)

File: clients/python/agentic-sandbox-client/k8s_agent_sandbox/async_k8s_helper.py (hunk 1 above)
Comment (chw120) [reply clients/python/agentic-sandbox-client/k8s_agent_sandbox/async_k8s_helper.py]: Imported validators directly at the top of the files.

File: clients/python/agentic-sandbox-client/k8s_agent_sandbox/k8s_helper.py (hunk 4 above)
Comment (chw120) [reply clients/python/agentic-sandbox-client/k8s_agent_sandbox/k8s_helper.py]: Imported validators directly at the top of the files.

File: clients/python/agentic-sandbox-client/k8s_agent_sandbox/async_k8s_helper.py (hunk 2 above)
Comment (chw120) [reply clients/python/agentic-sandbox-client/k8s_agent_sandbox/async_k8s_helper.py:329]: Updated to add the checks.

File: clients/python/agentic-sandbox-client/k8s_agent_sandbox/async_k8s_helper.py (hunk 3 above)
Comment (chw120) [reply clients/python/agentic-sandbox-client/k8s_agent_sandbox/async_k8s_helper.py:334]: Removed.

File: clients/python/agentic-sandbox-client/k8s_agent_sandbox/k8s_helper.py (hunk 5 above)
Comment (chw120) [reply clients/python/agentic-sandbox-client/k8s_agent_sandbox/k8s_helper.py:293]: Updated to add the checks.

File: clients/python/agentic-sandbox-client/k8s_agent_sandbox/k8s_helper.py (hunk 6 above)
Comment (chw120) [reply clients/python/agentic-sandbox-client/k8s_agent_sandbox/k8s_helper.py:298]: Removed.

File: clients/python/agentic-sandbox-client/k8s_agent_sandbox/utils.py (hunk 7 above)
Comment (chw120) [reply clients/python/agentic-sandbox-client/k8s_agent_sandbox/utils.py:123]: Added.

File: clients/python/agentic-sandbox-client/k8s_agent_sandbox/utils.py (hunk 8 above)
Comment (chw120) [reply clients/python/agentic-sandbox-client/k8s_agent_sandbox/utils.py:110]: Added a new check `_is_integer_label` for preventing it.

File: clients/python/agentic-sandbox-client/k8s_agent_sandbox/utils.py [hunk 9]
```diff
@@ -43,3 +43,52 @@ def construct_sandbox_claim_lifecycle_spec(shutdown_after_seconds: int) -> dict[
...
+def is_valid_ip(s: str) -> bool:
+    import ipaddress
+    try:
+        # Downstream callers construct URLs with `http://{ip_address}`,
+        # which only works for IPv4 literals without additional
+        # normalization. Reject IPv6 here to keep the returned value
+        # compatible with existing URL construction.
+        return isinstance(ipaddress.ip_address(s), ipaddress.IPv4Address)
+    except ValueError:
+        return False
```
[the remark this answers is not in the document]
Comment (chw120) [reply clients/python/agentic-sandbox-client/k8s_agent_sandbox/utils.py:109]: Updated in the new commit.

File: clients/python/agentic-sandbox-client/k8s_agent_sandbox/utils.py [hunk 10]
```diff
@@ -43,3 +43,56 @@ def construct_sandbox_claim_lifecycle_spec(shutdown_after_seconds: int) -> dict[
...
+
+
+def _is_integer_label(label: str) -> bool:
+    if not label:
+        return False
+    if label.isdigit():
+        return True
+    if label.lower().startswith("0x"):
```
Comment (aditya-shantanu) [inline clients/python/agentic-sandbox-client/k8s_agent_sandbox/utils.py:117]: Residual `0x` SSRF-bypass edge in `_is_integer_label`. A label of exactly `"0x"` (hex prefix, no digits) returns `False` here because `len(val) > 0` fails. That label is therefore neither treated as an integer label nor rejected by the hostname char-loop (`0`, `x` are both in the allowlist), so a value like `"0x.1"` or `"00.0x.0x.1"` passes `is_valid_gateway_hostname` and libc `inet_aton` coerces it to `0.0.0.1` (the `0.0.0.0/8` range). Confirmed locally:

```
is_valid_gateway_hostname("0x.1")        -> True
socket.gethostbyname("0x.1")             -> "0.0.0.1"
is_valid_gateway_hostname("00.0x.0x.1")  -> True  -> "0.0.0.1"
```

The broad integer-form bypass is fixed; this is the narrower leftover. Exploitability is limited (only `0.0.0.0/8` is reachable this way, not `127.0.0.1`), but since this validator is the security boundary for an untrusted Gateway status, a bare `"0x"` label should not pass. Suggest treating an empty-hex label as an integer label (or otherwise rejecting it), e.g. have `_is_integer_label` return `True` for a `"0x"`/`"0X"`-prefixed token regardless of remaining length, or reject any label that is `0x`-prefixed with no valid hex digits. Please apply locally rather than via the GitHub "Commit suggestion" button (the bot co-author breaks the CLA check).

**Review by aditya-shantanu — COMMENTED** [review]
Re-reviewed after the 2026-06-22 commits. The prior open feedback is addressed: validators moved to `utils.py` and imported at top; `isinstance(address, dict)` guard added in both helpers; the `/?#@` denylist removed in favor of pure allowlist; per-label 63-char limit and IPv6 rejection added; and my earlier integer-form SSRF bypass (e.g. `"2130706433"`) is now correctly rejected by `_is_integer_label`. Nice. One narrow residual gap below; everything else looks good.

File: clients/python/agentic-sandbox-client/k8s_agent_sandbox/utils.py (hunk 10 above)
Comment (chw120) [reply clients/python/agentic-sandbox-client/k8s_agent_sandbox/utils.py:117]: Fixed it in https://github.com/kubernetes-sigs/agent-sandbox/pull/756/commits/994dff70158debc16e61e529342a993b2efa8f37.

**Review by aditya-shantanu — APPROVED** [review]

File: clients/python/agentic-sandbox-client/k8s_agent_sandbox/k8s_helper.py [hunk 11]
```diff
@@ -289,9 +290,20 @@ def wait_for_gateway_ip(self, gateway_name: str, namespace: str, timeout: int) -
...
+                    for address in addresses:
+                        if not isinstance(address, dict):
+                            continue
+                        ip_address = address.get('value')
+                        if not ip_address:
+                            continue
+                        
+                        if not is_valid_ip(ip_address) and not is_valid_gateway_hostname(ip_address):
+                            logging.warning(
+                                "Gateway address rejected by validation: %r",
```
Comment (SHRUTI6991) [inline clients/python/agentic-sandbox-client/k8s_agent_sandbox/k8s_helper.py]: nit: throw a more intuitive error for users. 

`Gateway address rejected because ip_address is not valid`.
`Gateway address rejected because ip_address is not a valid gateway hostname`.

File: clients/python/agentic-sandbox-client/k8s_agent_sandbox/utils.py [hunk 12]
```diff
@@ -96,3 +96,56 @@ def select_pod_ip(ips: Sequence[object] | None) -> str | None:
...
+def is_valid_ip(s: str) -> bool:
+    if not isinstance(s, str):
+        return False
+    import ipaddress
+    try:
+        # Downstream callers construct URLs with `http://{ip_address}`,
+        # which only works for IPv4 literals without additional
+        # normalization. Reject IPv6 here to keep the returned value
+        # compatible with existing URL construction.
+        return isinstance(ipaddress.ip_address(s), ipaddress.IPv4Address)
```
Comment (SHRUTI6991) [inline clients/python/agentic-sandbox-client/k8s_agent_sandbox/utils.py]: we don't expect IPv6?

**Comment by SHRUTI6991** [comment]
Thanks for this change. It looks good. 

/lgtm

File: clients/python/agentic-sandbox-client/k8s_agent_sandbox/k8s_helper.py (hunk 11 above)
Comment (chw120) [reply clients/python/agentic-sandbox-client/k8s_agent_sandbox/k8s_helper.py]: Updated.

File: clients/python/agentic-sandbox-client/k8s_agent_sandbox/utils.py (hunk 12 above)
Comment (chw120) [reply clients/python/agentic-sandbox-client/k8s_agent_sandbox/utils.py]: Good catch. Found https://github.com/kubernetes-sigs/agent-sandbox/pull/1014 which supports/corrects IPv6 recently. Removed the comment.
